File: go/internal/sol/p0006/p0006.go
package p0006

import (
	"strconv"

	"pe-solver/internal/mylib"
)

func sumOfSquares(n int) int {
	return n * (n + 1) * (2*n + 1) / 6
}

func squareOfSum(n int) int {
	return mylib.Pow((n * (n + 1) / 2), 2)
}

func compute(limit int) string {
	if limit <= 0 {
		panic("The argument must be positive")
	}

	// The square of sum is equal or larger than the sum of squares.
	return strconv.FormatUint(
		uint64(squareOfSum(limit) - sumOfSquares(limit)),
		10,
	)
}

func Solve() string {
	return compute(100)
}

File: haskell/src/Sol/P0006.hs
module Sol.P0006 (compute, solve) where

sumOfSquares :: Int -> Int
sumOfSquares n = n * (n + 1) * (2 * n + 1) `div` 6

squareOfSum :: Int -> Int
squareOfSum n = (n * (n + 1) `div` 2) ^ (2 :: Int)

compute :: Int -> String
compute limit =
    show . abs $ sumOfSquares limit - squareOfSum limit

solve :: String
solve = compute 100
